core/src/main/java/io/grpc/internal/Channelz.java

```diff
@@ -449,22 +455,125 @@ public ChannelStats build() {
     }
   }
 
+  public interface SecurityInfoProvider {
```

Comment:
So to make sure I understand, `SecurityInfoProvider` isn't strictly necessary here, right? It isn't used with Channelz directly. It's mainly for convenience in Netty's negotiation. Everywhere else (including `getSecurityInfoProvider()`) seems it could handle `Security` directly.

I sort of expect this was done because of a comment of mine concerned about placing it in `Attributes`, where it is something like "always required, even if not used." That was more about matching the producer and consumer and it being clear when setting the Attribute is necessary (vs setting it all the time just imagining it may be used by some code, somewhere). Using explicit plumbing makes it easier to comprehend the data flow and knowing "when it is necessary" and the constraints around the value. There is benefit in explicit plumbing making it easier to communicate an internally-convenient object while also avoiding polluting the API with things like Providers and being able to tweak each piece separately. `Attributes` is convenient, but I really don't want to use it except when necessary, like communicating transport-specific information through io.grpc's API.

Reply:
Yes it is possible to just pass the `Security` around, but the reason this exists is to lazy-create the Security object only if it is needed. The toString() of a certificate looks like this, which felt a bit expensive to me:
```
[
[
  Version: V3
  Subject: CN=testclient, O=Internet Widgits Pty Ltd, ST=Some-State, C=AU
  Signature Algorithm: SHA256withRSA, OID = 1.2.840.113549.1.1.11

  Key:  Sun RSA public key, 1024 bits
  modulus: 165956023044986254443485235100948203493326685384458545615234492728849631380108078637198717973188982046560886128784222562463584010868957932010637062095083604661206388470176657085122939473572534129256975030377378114543169035021846486089803529987006058984879439854491857390215123100891877094523673506332331716589
  public exponent: 65537
  Validity: [From: Mon Nov 09 17:09:58 PST 2015,
               To: Thu Nov 06 17:09:58 PST 2025]
  Issuer: CN=testca, O=Internet Widgits Pty Ltd, ST=Some-State, C=AU
  SerialNumber: [    0a]

Certificate Extensions: 4
[1]: ObjectId: 2.5.29.35 Criticality=false
AuthorityKeyIdentifier [
[CN=testca, O=Internet Widgits Pty Ltd, ST=Some-State, C=AU]
SerialNumber: [    91c6191e 1d1a2a07]
]

[2]: ObjectId: 2.5.29.19 Criticality=false
BasicConstraints:[
  CA:false
  PathLen: undefined
]

[3]: ObjectId: 2.5.29.15 Criticality=false
KeyUsage [
  DigitalSignature
  Non_repudiation
  Key_Encipherment
]

[4]: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 00 75 B5 B9 56 69 7F 0A   76 30 A9 D3 F7 98 E1 C3  .u..Vi..v0......
0010: 01 15 3E F3                                        ..>.
]
]

Unparseable certificate extensions: 2
[1]: ObjectId: 2.5.29.17 Criticality=false
Unparseable SubjectAlternativeName extension due to
java.io.IOException: No data available in passed DER encoded value.

0000: 30 00                                              0.

[2]: ObjectId: 2.5.29.18 Criticality=false
Unparseable IssuerAlternativeName extension due to
java.io.IOException: No data available in passed DER encoded value.

0000: 30 00                                              0.

]
  Algorithm: [SHA256withRSA]
  Signature:
0000: 7F 83 0C 93 EB 1D CD DE   F6 D0 37 EB 43 43 C5 DA  ..........7.CC..
0010: 00 DA B9 1D CC F6 E0 6E   6E 8C C3 B8 CB 85 EA E7  .......nn.......
0020: C0 B9 00 49 48 A1 90 10   68 9B 5D B5 1E 96 BC 07  ...IH...h.].....
0030: 9A 7A A5 52 34 F2 7E 89   55 73 11 01 55 A1 78 54  .z.R4...Us..U.xT
0040: 51 E6 9F BB BA BC FE D2   D4 D7 B1 20 37 4D 50 77  Q.......... 7MPw
0050: 14 2E AC 48 6F B4 F3 97   C2 58 AB C3 8E 1C 3D AB  ...Ho....X....=.
0060: 43 2E 07 B3 C0 57 69 52   33 C3 86 1C 68 21 80 31  C....WiR3...h!.1
0070: 2D 22 0D 7D 39 D9 2C BB   7B 62 61 3F C3 68 C6 FA  -"..9.,..ba?.h..

]```